### DISCUSSION TITLE
How or where to update output parameters after execution

### DISCUSSION THREAD

**Discussion by wanaxe** [linq2db/linq2db#3593] [Q&A] [answered]
For non query or stored procedure commands, we need to update output parameters after query execution. For v3.6.0, the code was like

```c#
public async Task<int> ExecuteNonQueryAsync(string sql, params DataParameter[] dataParameters)
{
	using var dataContext = CreateDataConnection();
	var command = new CommandInfo(dataContext, sql, dataParameters);
	var affectedRecords = await command.ExecuteAsync();

	// update output parameters
	foreach (var dataParam in dataParameters.Where(p => p.Direction == ParameterDirection.Output))
	{
		if (dataConnection.Command is IDbCommand command 
			&& command.Parameters.Count > 0 
			&& command.Parameters.Contains(dataParam.Name) 
			&& command.Parameters[dataParam.Name] is IDbDataParameter param)
		{
			dataParam.Value = param.Value;
		}
	}
	
	return affectedRecords;
}
```

Now `DataConnection.Command` is gone since v4.0.0.  How and where (e.g., which `Interceptor` or event) to do the similar thing?

Please kindly advise.

**Comment by MaceWindu** [comment, marked as answer]
https://linq2db.github.io/articles/general/interceptors.html#icommandinterceptor

You can try to call `ExecuteNonQueryAsync` in `ICommandInterceptor.ExecuteNonQueryAsync` and configure parameters there.

**Reply by wanaxe** [reply]
I saw `ICommandInterceptor.ExecuteNonQueryAsync` is "triggered before `ExecuteNonQuery/ExecuteNonQueryAsync` call on command". Does it mean it is triggered before the query execution?

We need to update output parameters after query execution.

**Reply by MaceWindu** [reply]
Yes, but you can call execute method here and return result. In this case linq2db will use it instead of invoking execute

**Reply by wanaxe** [reply]
Got it. Thanks a lot!

How about `QueryProc<T>`? I cannot find a similar interceptor event.

**Reply by MaceWindu** [reply]
`QueryProc` case is more tricky. It calls `ExecuteReader[Async]` and main issue here is that output parameters (at least for SQL Server, cannot tell about other providers right now, but probably the same) available only after all data consumed from reader. E.g. this is the reason why we generate `ToList` calls in scaffolded context for procedures with output parameters.

I don't see currently how you can handle it with interceptors, probably we need another interceptor to trigger after data reader consumed.

**Reply by wanaxe** [reply]
Understood, thank you so much! Hopefully that intercaptor could be added in future realease.

**Reply by MaceWindu** [reply]
New interceptors will be released next week https://github.com/linq2db/linq2db/pull/3724

**Reply by wanaxe** [reply]
Great!!!
